federation/server.c

```diff
@@ -0,0 +1,765 @@
+    spiffebundle_EndpointServer *server, const char *path,
+    spiffebundle_Source *new_source, spiffeid_TrustDomain td)
+{
+    if(!server) {
+        return ERR_NULL;
+    }
+    if(!path) {
+        return ERR_BAD_ARGUMENT;
+    }
+    if(!new_source) {
+        return ERR_NULL_DATA;
+    }
+    if(!td.name) {
+        return ERR_INVALID_TRUSTDOMAIN;
+    }
+
+    mtx_lock(&server->mutex);
+    {
+        int idx = shgeti(server->bundle_sources, path);
+        if(idx < 0) { // not found
+            mtx_unlock(&server->mutex);
+            return ERR_NOT_FOUND;
+        }
+        server->bundle_sources[idx].value = new_source;
+        if(server->bundle_tds[idx].value) {
+            arrfree(server->bundle_tds[idx].value);
+        }
+        server->bundle_tds[idx].value = string_new(td.name);
+    }
+    mtx_unlock(&server->mutex);
+    return NO_ERROR;
+}
+
+// removes bundle from server.
+err_t spiffebundle_EndpointServer_RemoveBundle(
+    spiffebundle_EndpointServer *server, const char *path)
+{
+    if(!server) {
+        return ERR_NULL;
+    }
+    if(!path) {
+        return ERR_BAD_ARGUMENT;
+    }
+
+    mtx_lock(&server->mutex);
+    {
+        int idx = shgeti(server->bundle_sources, path);
+        if(idx < 0) { // not found
+            mtx_unlock(&server->mutex);
+            return ERR_NOT_FOUND;
+        }
+
+        // free string from td strings map
+        arrfree(server->bundle_tds[idx].value);
+
+        shdel(server->bundle_sources, path);
+        shdel(server->bundle_tds, path);
+    }
+    mtx_unlock(&server->mutex);
+
+    return NO_ERROR;
+}
+
+// load keys to use with 'https_web'
+// register a HTTPS_WEB endpoint, for starting with
+// spiffebundle_EndpointServer_ServeEndpoint
+spiffebundle_EndpointInfo *spiffebundle_EndpointServer_AddHttpsWebEndpoint(
+    spiffebundle_EndpointServer *server, const char *base_url, X509 **certs,
+    EVP_PKEY *priv_key, err_t *error)
+{
+    if(!server) {
+        *error = ERR_NULL;
+        return NULL;
+    }
+    if(!base_url) {
+        *error = ERR_BAD_ARGUMENT;
+        return NULL;
+    }
+    if(!certs || !(certs[0])) {
+        *error = ERR_CERTIFICATE_VALIDATION;
+        return NULL;
+    }
+    priv_key = x509svid_validatePrivateKey(priv_key, certs[0], error);
+    if(!priv_key) {
+        *error = ERR_PRIVKEY_VALIDATION;
+        return NULL;
+    }
+
+    mtx_lock(&server->mutex);
+    int idx = shgeti(server->endpoints, base_url);
+    if(idx >= 0) {
+        EVP_PKEY_free(priv_key);
+        mtx_unlock(&server->mutex);
+        *error = ERR_EXISTS;
+        return NULL;
+    }
+
+    spiffebundle_EndpointInfo *e_info = spiffebundle_EndpointInfo_New();
+    mtx_lock(&e_info->mutex);
+    shput(server->endpoints, base_url, e_info);
+    e_info->server = server;
+    mtx_unlock(&server->mutex);
+
+    // create svid with blank spiffeid for holding the key and certificates
+    x509svid_SVID *svid = calloc(1, sizeof *svid);
```

Comment:
You could use the 'x509svid_newSVID' function here.

Reply:
see below